### PR TITLE
Use PRECIS and IDNA2008 instead of stringprep

### PR DESCRIPTION
**Pull request by SamWhited**
This makes JIDs compatible with [RFC 7622](https://tools.ietf.org/html/rfc7622) by replacing the deprecated stringprep with PRECIS ([RFC 7564](https://tools.ietf.org/html/rfc7564)). This was a quick find-replace and almost certainly needs looking over / testing by someone more familiar with the codebase.

This adds two new dependencies (unfortunately):

* [idna](https://pypi.python.org/pypi/idna) (since the built in version of IDNA only supports IDNA2003)
* [precis-i18n](https://pypi.python.org/pypi/precis-i18n/0.2.2)

I left the `prepfunc` kwarg on the String type alone, but it may also be worth changing that API; maybe to "codec" to make it more generic? Or it may not be worth messing with.

### CODE REVIEW

**Comment by horazont** [comment]
That’s great! Last time I checked there was no precis implementation for Python. I’ll take a look at this tomorrow.

**Comment by SamWhited** [comment]
It's brand new (as of a few days ago, I think); so far it passes the test vectors RFC I'm working on with flying colors though, so it seems like it's off to a good start.

File: aioxmpp/forms/xso.py [hunk 1]
```diff
@@ -145,9 +145,9 @@ class FieldType(enum.Enum):
 
           The field enables an entity to gather or provide multiple Jabber IDs.
           Each provided JID SHOULD be unique (as determined by comparison that
-          includes application of the Nodeprep, Nameprep, and Resourceprep
-          profiles of Stringprep as specified in XMPP Core), and duplicate JIDs
-          MUST be ignored.
+          includes application of the Username Case Mapped, and Nickname
+          profiles of PRECIS or IDNA2008 as specified in RFC 7622), and duplicate
```
Comment (horazont) [inline aioxmpp/forms/xso.py]: Use ``[…] specified in:rfc:`7622`)[…]`` to generate a link to the RFC.

File: aioxmpp/muc/xso.py [hunk 2]
```diff
@@ -180,7 +180,7 @@ class ActorBase(xso.XSO):
 
     nick = xso.Attr(
         "nick",
-        type_=xso.String(aioxmpp.stringprep.resourceprep),
+        type_=xso.String(precis.Nickname.enforce),
```
Comment (horazont) [inline aioxmpp/muc/xso.py]: This raises AttributeError here:

```
>>> import aioxmpp.muc
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/horazont/Projects/python/aioxmpp-sam/aioxmpp/muc/__init__.py", line 100, in <module>
    from .service import Service, Occupant, Room  # NOQA
  File "/home/horazont/Projects/python/aioxmpp-sam/aioxmpp/muc/service.py", line 35, in <module>
    from . import xso as muc_xso
  File "/home/horazont/Projects/python/aioxmpp-sam/aioxmpp/muc/xso.py", line 174, in <module>
    class ActorBase(xso.XSO):
  File "/home/horazont/Projects/python/aioxmpp-sam/aioxmpp/muc/xso.py", line 183, in ActorBase
    type_=xso.String(precis.Nickname.enforce),
AttributeError: module 'precis_i18n' has no attribute 'Nickname'
```

File: aioxmpp/muc/xso.py [hunk 3]
```diff
@@ -208,7 +208,7 @@ class ItemBase(xso.XSO):
...
```
Comment (horazont) [inline aioxmpp/muc/xso.py]: I guess this has the same problem as above.

**Comment by horazont** [comment]
If you have trouble getting into working with the code, feel free to re-push after fixing the ``encode`` vs. ``enforce`` thing I commented on. That should already reduce the breakage significantly.

**Comment by horazont** [comment]
FWIW, there seems to be some [discussion on including IDNA2008 in Python 3.x](https://bugs.python.org/issue17305).

Unfortunately, it’s "contributions welcome", and no-one has stepped up yet.

File: aioxmpp/structs.py [hunk 4]
```diff
@@ -633,20 +634,11 @@ class JID(collections.namedtuple("JID", ["localpart", "domain", "resource"])):
 
     def __new__(cls, localpart, domain, resource, *, strict=True):
         if localpart:
-            localpart = nodeprep(
-                localpart,
-                allow_unassigned=not strict
-            )
+            localpart = localpart.encode('UsernameCaseMapped')
```
Comment (SamWhited) [inline aioxmpp/structs.py]: No particular reason, I was just using this as an excuse to play around with the new PRECIS library a bit and didn't really think through the changes.

**Comment by SamWhited** [comment]
Closing this for now; looks like there's a lot wrong with it and I don't have any significant time to invest in this. I may resume it at a later date.

File: aioxmpp/forms/xso.py (hunk 1 above)
Comment (SamWhited) [inline aioxmpp/forms/xso.py]: Oh nifty, TIL. Thanks for the pointer.

**Comment by horazont** [comment]
No worries. It’s already valuable to know that there’s now a PRECIS implementation for Python out there. Pinging #9 for future reference.

File: aioxmpp/muc/xso.py (hunk 2 above)
Comment (SamWhited) [inline aioxmpp/muc/xso.py]: Fixed in branch (this API was changed with a release yesterday and the top level constants were removed).

**Pull request closed without merging**